### PR TITLE
Add support for error_on_eviction job config field.

### LINKED ISSUE #10188: Prow job started twice
**What happened**:
The ci-kubernetes-e2e-gce-scale-performance job was started two times on 2018-11-16: at 21:40 CET, and 9:02 CET.

**What you expected to happen**:
Only single run to happen, the one at 9:02 CET which is specified in the job config.

**How to reproduce it (as minimally and precisely as possible)**:

**Please provide links to example occurrences, if any**:

https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/258
https://gubernator.k8s.io/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-performance/259

**Anything else we need to know?**:

/cc @wojtek-t 

/kind bug

### PR DESCRIPTION
**Pull request by cjwagner**
This PR adds a per job config field, `error_on_eviction` that changes how `plank` handles evicted pods for the job. If this field is set to `true`, instead of silently creating a new pod for the ProwJob to continue with, `plank` will mark the ProwJob as completed and errored.

I also included some other fixes for problems that I noticed along the way:
- Move `Cluster` field defaulting to be with the rest of the config defaulting.
- Actually transfer `Namespace` and `BuildSpec` fields from job config to ProwJobSpec (previously they were ignored when creating new ProwJobSpecs).
- Generalize the steps of ProwJobSpec creation that are common to the different job types.
- Transfer `Cluster` field from job config to ProwJobSpec for all agents, not just the kubernetes agent (build-CRD controller will need this).

fixes #10188 
/cc @fejta @BenTheElder @mborsz @stevekuznetsov 
/kind feature
/kind bug

### CODE REVIEW

File: prow/config/config_test.go [hunk 1]
```diff
@@ -90,6 +91,15 @@ func TestDefaultJobBase(t *testing.T) {
 				j.Namespace = &p
 			},
 		},
+		{
+			name: "empty cluster becomes DefaultClusterAlias",
```
Comment (stevekuznetsov) [inline prow/config/config_test.go]: Do we have a test for `error_on_eviction` with non-kubernetes agent?

**Comment by BenTheElder** [comment]
Ultimate bikeshed: are we settled on foo_bar vs fooBar vs ... for prow config?

**Comment by stevekuznetsov** [comment]
If we ever want to converge on k8s-style we need `fooBar`

**Comment by cjwagner** [comment]
> Ultimate bikeshed: are we settled on foo_bar vs fooBar vs ... for prow config?

I used underscores for consistency with neighboring fields. I personally prefer `fooBar`.
This is a widespread problem though and fixing it will be a breaking change, so I think we should fix all of the fields at once (and provide a migration script) if we are going to make that change.

File: prow/config/config_test.go (hunk 1 above)
Comment (cjwagner) [reply prow/config/config_test.go]: We do now.

**Comment by BenTheElder** [comment]
Seems like we ought to make new fields follow our desired convention so we have fewer to change. We also ought to spell this out somewhere in a development doc.

**Comment by BenTheElder** [comment]
(also a +1 for fooBar and following k8s style...)

**Comment by stevekuznetsov** [comment]
I am ok with adding `foo_bar` and a migration (or a shim loader that accepts `foo_bar` or `fooBar` for some time on the migration)

**Comment by cjwagner** [comment]
PTAL

**Comment by krzyzacy** [comment]
cc @wojtek-t

File: prow/test/data/kubernetes-defaulted-decoration-prow-job.yaml [hunk 2]
```diff
@@ -31,6 +31,7 @@ spec:
       initupload: initupload:default
       sidecar: sidecar:default
   job: kubernetes-defaulted-decoration
+  namespace: default
```
Comment (fejta) [inline prow/test/data/kubernetes-defaulted-decoration-prow-job.yaml]: What are these about? This seems wrong?

File: prow/config/config.go [hunk 3]
```diff
@@ -1124,6 +1126,9 @@ func (c *ProwConfig) defaultJobBase(base *JobBase) {
 		s := c.PodNamespace
 		base.Namespace = &s
 	}
+	if base.Cluster == "" {
```
Comment (fejta) [inline prow/config/config.go]: why is this necessary?

File: prow/plank/controller.go [hunk 4]
```diff
@@ -374,8 +374,17 @@ func (c *Controller) syncPendingJob(pj kube.ProwJob, pm map[string]kube.Pod, rep
 
 		case kube.PodFailed:
 			if pod.Status.Reason == kube.Evicted {
+				// Pod was evicted.
+				if pj.Spec.ErrorOnEviction {
```
Comment (fejta) [inline prow/plank/controller.go]: I'm a bit confused why this isn't the only change?

**Comment by cjwagner** [comment]
@fejta I think the PR body text answers your questions. Let me know if it doesn't.

**Comment by BenTheElder** [comment]
sounds fair enough. I wonder if maybe we shouldn't bite the bullet and make that change soon though.

**Comment by krzyzacy** [comment]
@wojtek-t if there's at most one job gonna be running in the project at any time, we probably can do a clean up before each test?

**Comment by cjwagner** [comment]
> Yes - we definitely want (not for every job, but for those that are running in predefined projects yes).

Which jobs are these?

> What I'm wondering about (not in this PR, but maybe as a follow up) is whether we should somehow try to clean up the project in case of eviction.

I think Boskos can handle this? @krzyzacy 

Lets make config changes to enable this feature on jobs in separate PRs.

**Comment by krzyzacy** [comment]
> Which jobs are these?

Scalability jobs.

> I think Boskos can handle this? @krzyzacy

The large scalability jobs are not using boskos yet, we only have one project and scheduling them using cron feature as periodic jobs. A custom pre-test hook in kubetest can probably handle this?

**Comment by cjwagner** [comment]
I understand that we're talking about scalability jobs. @wojtek-t said not for every job though, only the ones running in predefined projects. So I am wondering which specific scalability jobs that is referring to.

Cleaning up when the job starts sounds like a good idea to me. (Might also be nice for debugging completed jobs.)

**Comment by krzyzacy** [comment]
there are 7 jobs bind to `kubernetes-scale` project in https://github.com/kubernetes/test-infra/blob/92f12c1fc552f18182f5c03c24043ff8579150c9/config/jobs/kubernetes/sig-scalability/sig-scalability-periodic-jobs.yaml

**Comment by BenTheElder** [comment]
cc @amwat for some more jobs ^

**Comment by wojtek-t** [comment]
> @wojtek-t ah makes sense. The issue here is if the pod is evicted then we won't be able to perform any actions defined in kubetest. 

Yeah I know that.
Just to clarify - my concern here wasn't really correctness (which I think we have it roughly working), but rather cost of it.

**Comment by wojtek-t** [comment]
They are not - we can't have O(10) project with such a high quota. Given that there is one at a time running, there will be exactly one project.

**Comment by krzyzacy** [comment]
good point ^^

**Comment by wojtek-t** [comment]
Ohh - so what you're suggesting is making a group consisting of only this single project and use boskos that will be taking from that group with single project. That should work, I guess.

**Comment by krzyzacy** [comment]
@wojtek-t currently janitor will wipe out the entire project after each run, will there be none-CI clusters running in the scale project?

**Comment by wojtek-t** [comment]
> will there be none-CI clusters running in the scale project?

No - there is a separate dedicated project for all ad-hoc/manual/etc. testing.
We can safely assume that the kubernetes-scale project is for CI only.

**Comment by krzyzacy** [comment]
sgtm, I'll take a stab before next release starts :-) cc @spiffxp @BenTheElder

**Comment by BenTheElder** [comment]
Yes exactly, boskos is pretty good and solving the "botched tests leave around resources" issue, so a group of a single project should work. As a side benefit we have some monitoring / metrics for this. http://velodrome.k8s.io/dashboard/db/boskos-dashboard?orgId=1

**Comment by BenTheElder** [comment]
Thanks @krzyzacy !
